cpp/src/arrow/device.h

```diff
@@ -363,4 +363,22 @@ class ARROW_EXPORT CPUMemoryManager : public MemoryManager {
 ARROW_EXPORT
 std::shared_ptr<MemoryManager> default_cpu_memory_manager();
 
+/// \brief Copy all buffers of an array to destination MemoryManager
+///
+/// This utilizes MemoryManager::CopyBuffer to create a new Array recursively copying
+/// all buffers, and all children buffers, to the destination MemoryManager. This
+/// includes any dictionaries if applicable.
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> CopyArrayTo(const Array& array,
```

Comment:
Nit, but I think this could simply be a member of Array.

Also, do we want a similar CopyOrView? If the array, or part thereof, is already located on the right device, we probably don't want to incur any pointless memory copy.

Reply:
I can make those changes, sure

Reply:
`CopyOrView` creates weird expectations about the returned values: if you copy, the caller can have unique ownership of the array, if it's a view, you have to share. Another possible solution to the problem @pitrou brought up is failing the operation if source and destination are the same device, forcing the caller to be aware of data location before it calls the copy function. The compromise solution: provide `CopyOrView` in terms of a `CopyArrayTo` that returns a bad status when called for arrays on the same device.

(I understand that the memory model in Arrow is that every buffer is immutable and always shared, but we might want these rules to be relaxed at some point.)

Reply:
If the caller *requires* unique ownership of the array, then they should just call `CopyTo` in the first place. 

We don't want to fail if the source and destination are the same device because maybe they are just intentionally doing a full copy to get unique ownership, but also because the device is tracked on a per-buffer basis, not at the array level. Having a `CopyOrView` allows the user to attempt to avoid copies if they don't need unique ownership of the buffers, so i don't think it's weird expectations about the returned values. It's very clear what the expectations are:

1. If you need unique ownership, call `Copy`
2. If you don't need unique ownership and want to avoid copies if possible then use `ViewOrCopy` (for example if you're using Cuda Host memory you can get a CPU accessible view without having to manually perform a copy)

> This might be a bad idea without a guarantee that all buffers making up an array or RecordBatch are on the same device. :/

One way to ensure they are all on the same device is to call this function! :smile:

Reply:
You're right. I'm thinking about ownership in a weird way (at least relative to how Arrow works).